# 8 Word Embeddings

Documentation:

* https://fasttext.cc/docs/en/english-vectors.html
* https://fasttext.cc/docs/en/unsupervised-tutorial.html
* https://fasttext.cc/docs/en/html/index.html

1. Install fastText and load the pre-trained word embeddings into a fastText model.

In [1]:
!pip -q install fasttext

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.4/73.4 kB 1.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done


In [2]:
import fasttext
import fasttext.util
fasttext.util.download_model('en', if_exists='ignore')  # English

'cc.en.300.bin'

In [3]:
# Task 8.2
print ("Loading the model")

# Load the model
model = fasttext.load_model('cc.en.300.bin')

Loading the model


2\. How large is the vocabulary?

In [13]:
print("The vocabulary size is", len(model.words), "words")


The vocabulary size is 2000000 words


3\. How many dimensions does the model have?

In [14]:
print("The model have", model.get_dimension(), "dimensions")

The model have 300 dimensions


4\. Print the word vector for "Munich".

In [19]:
print("The word vector for Munichis:")
print(model.get_word_vector('Munich'))

The word vector for Munichis:
[-0.01772507  0.04481639 -0.0458482  -0.01564159  0.00569844 -0.01493978
 -0.00475513  0.05498182  0.03653041  0.05107911 -0.00792894 -0.04519449
  0.02314676 -0.04049589  0.0533328  -0.01538926 -0.07663719 -0.1266364
 -0.0860248   0.01669377 -0.06406741 -0.01205093  0.10280135 -0.05018425
 -0.00066486 -0.08689803 -0.03842681 -0.08041406  0.16761142  0.11794221
  0.01121775  0.00119475 -0.01672026  0.01051223 -0.08705565  0.01279867
 -0.04793347  0.07529095  0.06565578 -0.03357543  0.06183482 -0.01693818
  0.07336422 -0.12507132 -0.06156905  0.01393662 -0.02854991  0.04012572
  0.08620507  0.07610758  0.00872343  0.12036866 -0.09645183 -0.0086469
 -0.04031846 -0.03134852 -0.08990994  0.01966693  0.11320422 -0.08001418
  0.028819   -0.01371382  0.06330752  0.09222437  0.02637078 -0.05378804
 -0.09224798 -0.05626002  0.02702164  0.03305807 -0.01610627  0.09820458
 -0.0245055  -0.12223028 -0.06889763 -0.00073068  0.09101934 -0.07081737
 -0.0151146   0.0063146

5\. What are the most similar terms to "LRZ"?

In [11]:
model.get_nearest_neighbors("LRZ")

[(0.6676446795463562, 'SuperMUC'),
 (0.6408230066299438, 'WLCG'),
 (0.6323488354682922, 'FZI'),
 (0.6296346783638, 'PDSF'),
 (0.624134361743927, 'FZK'),
 (0.6112837195396423, 'ESSL'),
 (0.6090673208236694, 'iVEC'),
 (0.6009016633033752, 'ETHZ'),
 (0.6007493734359741, 'XK7'),
 (0.5999080538749695, 'NERSC')]

6\. Find the analogy: Berlin is to Germany as ??? is to Denmark.

In [12]:
model.get_analogies("Berlin", "Germany", "Denmark")

[(0.7887206673622131, 'Copenhagen'),
 (0.7260220050811768, 'Aarhus'),
 (0.6744620203971863, 'Stockholm'),
 (0.6726727485656738, 'Århus'),
 (0.6658076047897339, 'Frederiksberg'),
 (0.642078697681427, 'Odense'),
 (0.6269904375076294, 'Christiania'),
 (0.6266441345214844, 'Roskilde'),
 (0.6243226528167725, 'Oslo'),
 (0.6190618276596069, 'Copenhagan')]

In [21]:
vec1 = model.get_word_vector("Munich")
vec2 = model.get_word_vector("Berlin")

similarity = np.dot(vec1, vec2) / (np.linalg.norm(vec1) * np.linalg.norm(vec2))
print(similarity)

0.7393219


In [34]:
vec1 = model.get_word_vector("Munich1")
vec2 = model.get_word_vector("Berlin")

similarity = np.dot(vec1, vec2) / (np.linalg.norm(vec1) * np.linalg.norm(vec2))
print(similarity)

0.22931011


In [23]:
vec1 = model.get_word_vector("Munish")
vec2 = model.get_word_vector("Berlin")

similarity = np.dot(vec1, vec2) / (np.linalg.norm(vec1) * np.linalg.norm(vec2))
print(similarity)

0.14827582


In [27]:
vec1 = model.get_word_vector("Munih")
vec2 = model.get_word_vector("Berlin")

similarity = np.dot(vec1, vec2) / (np.linalg.norm(vec1) * np.linalg.norm(vec2))
print(similarity)

0.017680673


In [28]:
vec1 = model.get_word_vector("Munivh")
vec2 = model.get_word_vector("Berlin")

similarity = np.dot(vec1, vec2) / (np.linalg.norm(vec1) * np.linalg.norm(vec2))
print(similarity)

0.14676662


I was curious to investigate how similarity works and how the model handles typos by replacing a character with one close on the keyboard, adding a letter, or missing one.
The results were surprising; even a single small typo caused the similarity to drop significantly:

- Munich  vs Berlin → 0.739  (correct spelling, high similarity)
- Munich1 vs Berlin → 0.229  (added a number -> similarity collapsed)
- Munish  vs Berlin → 0.148  (one wrong character —> already far from the original)
- Munih   vs Berlin → 0.018  (missing one character —> nearly unrecognizable)
- Munivh  vs Berlin → 0.147  (keyboard-close typo —> similar drop to Munish)

In [36]:
model.get_nearest_neighbors("Munich")


[(0.7680739164352417, 'Düsseldorf'),
 (0.7581844329833984, 'München'),
 (0.7536050081253052, 'Münich'),
 (0.7527161836624146, 'Stuttgart'),
 (0.7509379982948303, 'Munchen'),
 (0.7405064105987549, 'Dusseldorf'),
 (0.7393224239349365, 'Berlin'),
 (0.7358995079994202, 'Leipzig'),
 (0.7302831411361694, 'Hamburg'),
 (0.7237011790275574, 'Frankfurt')]

In [38]:
model.get_nearest_neighbors("Munish")

[(0.76488196849823, 'Saurabh'),
 (0.7448103427886963, 'Sandeep'),
 (0.7378804087638855, 'Neeraj'),
 (0.7337072491645813, 'Anuj'),
 (0.730432391166687, 'Rakesh'),
 (0.7278562784194946, 'Avnish'),
 (0.7260376811027527, 'Sumit'),
 (0.72511887550354, 'Dheeraj'),
 (0.7197767496109009, 'Siddharth'),
 (0.7176060080528259, 'Ajay')]

In [40]:
model.get_nearest_neighbors("munich1")

[(0.6520166397094727,
  'GermanyBerlinBremenCologneDortmundDresdenDusseldorfFrankfurtFriedrichshafenHamburgHanoverKarlsruhe'),
 (0.6258032321929932,
  'CITIESAgartalaAgraAhmedabadAllahabadAmritsarAurangabadBareillyBhopalBhubaneswarChandigarhCoimbatoreCuttackDehradunErodeFaridabadGoaGurgaonGuwahatiHubliImphalIndoreJaipurJammuJamshedpurJindKanpurKochiKolhapurKozhikodeLucknowLudhianaMaduraiMangaloreMeerutMysoreNagpurNashikNavi'),
 (0.6251739859580994,
  'IrelandConnachtGalwayLeitrimMayoRoscommonSligoLeinsterCarlowCo'),
 (0.6244726181030273,
  'MumbaiBeedDhuleWardhaRaigarhRatnagiriLaturYavatmalChandrapurNandedAkolaImphalShillongAizawlDimapurBhubaneswarCuttackPondicherryPatialaMohaliLudhianaAmritsarJalandharBathindaHoshiarpurSangrurGurdaspurMogaFirozpurFatehgarh'),
 (0.6187490224838257,
  'SahibKapurthalaJaipurUdaipurAlwarJodhpurAjmerSikarKotaGanganagarBikanerJhunjhunuHanumangarhGangtokCoimbatoreNamakkalKanyakumariSalemMaduraiVelloreKanchipuramErodeTirunelveliThanjavurVillupuramDindigulTric

It even affected the neighboring words significantly!

Munich  → neighbors are German cities: Düsseldorf, München, Stuttgart, Berlin, Leipzig, Hamburg, Frankfurt (True)

Munish  → neighbors are completely different.

Munich1 → neighbors are garbage non sense strings.

This made me think: how do modern models like the latest ChatGPT or Claude handle typos so well?  
Is it because the matrices/embeddings or the Vocab are way bigger? Or is this the sub-word tokenization role? Or are they trained on on large, noisy datasets containing errors that allow these models to recognize typos and erros?


7\. Study the Transformer notebook `Transformers-pipelines.ipynb` and investigate the language tasks that can be supported by transformer models.